ivy: cdkVirtualForOf template type errors

#### Reproduction
Build dev-app with Ivy 
 
#### Expected Behavior
The demo compiles.

#### Actual Behavior
```ts
ERROR: /Users/splaktar/Git/angular/material2/src/dev-app/virtual-scroll/BUILD.bazel:5:1: Couldn't build file src/dev-app/virtual-scroll/virtual-scroll-demo-module.ngfactory.js: Compiling Angular templates (ngtsc - devmode) //src/dev-app/virtual-scroll:virtual-scroll failed (Exit 1)
src/dev-app/virtual-scroll/virtual-scroll-demo.html(90,8): error TS2322: Type 'number[] | null' is not assignable to type 'number[] | DataSource<number> | Observable<number[]> | Iterable<number> | undefined'.
  Type 'null' is not assignable to type 'number[] | DataSource<number> | Observable<number[]> | Iterable<number> | undefined'.
```

A fix similar to what was done for `ngForOf` in PR https://github.com/angular/angular/pull/31371 is also needed for `cdkVirtualForOf`.

#### Workaround
Change
```html
<div *cdkVirtualFor="let size of observableData | async; let i = index" class="demo-item [style.height.px]="size">
```
to
```html
<div *cdkVirtualFor="let size of (observableData | async)!; let i = index" class="demo-item" [style.height.px]="size">
```

#### Environment

- Angular: 9.0.0-next.10
- CDK/Material: `master`
- Browser(s): N/A
- Operating System (e.g. Windows, macOS, Ubuntu): macOS Mojave

--- a/src/cdk/scrolling/virtual-for-of.ts
+++ b/src/cdk/scrolling/virtual-for-of.ts
@@ -83,10 +83,10 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy 
 
   /** The DataSource to display. */
   @Input()
-  get cdkVirtualForOf(): DataSource<T> | Observable<T[]> | NgIterable<T> {
+  get cdkVirtualForOf(): DataSource<T> | Observable<T[]> | NgIterable<T> | null | undefined {
     return this._cdkVirtualForOf;
   }
-  set cdkVirtualForOf(value: DataSource<T> | Observable<T[]> | NgIterable<T>) {
+  set cdkVirtualForOf(value: DataSource<T> | Observable<T[]> | NgIterable<T> | null | undefined) {
     this._cdkVirtualForOf = value;
     const ds = isDataSource(value) ? value :
         // Slice the value if its an NgIterable to ensure we're working with an array.
@@ -94,7 +94,7 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy 
             value instanceof Observable ? value : Array.prototype.slice.call(value || []));
     this._dataSourceChanges.next(ds);
   }
-  _cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T>;
+  _cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T> | null | undefined;
 
   /**
    * The `TrackByFunction` to use for tracking changes. The `TrackByFunction` takes the index and
@@ -363,7 +363,9 @@ export class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy 
     // comment node which can throw off the move when it's being repeated for all items.
     return this._viewContainerRef.createEmbeddedView(this._template, {
       $implicit: null!,
-      cdkVirtualForOf: this._cdkVirtualForOf,
+      // It's guaranteed that the iterable is not "undefined" or "null" because we only
+      // generate views for elements if the "cdkVirtualForOf" iterable has elements.
+      cdkVirtualForOf: this._cdkVirtualForOf!,
       index: -1,
       count: -1,
       first: false,

--- a/tools/public_api_guard/cdk/scrolling.d.ts
+++ b/tools/public_api_guard/cdk/scrolling.d.ts
@@ -60,8 +60,8 @@ export declare class CdkScrollable implements OnInit, OnDestroy {
 }
 
 export declare class CdkVirtualForOf<T> implements CollectionViewer, DoCheck, OnDestroy {
-    _cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T>;
-    cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T>;
+    _cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T> | null | undefined;
+    cdkVirtualForOf: DataSource<T> | Observable<T[]> | NgIterable<T> | null | undefined;
     cdkVirtualForTemplate: TemplateRef<CdkVirtualForOfContext<T>>;
     cdkVirtualForTemplateCacheSize: number;
     cdkVirtualForTrackBy: TrackByFunction<T> | undefined;

